crates/ra_proc_macro/src/process.rs

```diff
@@ -0,0 +1,209 @@
+//! Handle process life-time and message passing for proc-macro client
+
+use crossbeam_channel::{bounded, Receiver, Sender};
+use ra_tt::Subtree;
+
+use crate::msg::{ErrorCode, Message, Request, Response, ResponseError};
+use crate::rpc::{ExpansionResult, ExpansionTask, ListMacrosResult, ListMacrosTask, ProcMacroKind};
+
+use io::{BufRead, BufReader};
+use std::{
+    convert::{TryFrom, TryInto},
+    io::{self, Write},
+    path::{Path, PathBuf},
+    process::{Child, Command, Stdio},
+};
+
+#[derive(Debug, Default)]
+pub(crate) struct ProcMacroProcessSrv {
+    inner: Option<Sender<Task>>,
+}
+
+#[derive(Debug)]
+pub(crate) struct ProcMacroProcessThread {
+    // XXX: drop order is significant
+    sender: SenderGuard,
+    handle: jod_thread::JoinHandle<()>,
+}
+
+#[derive(Debug)]
+struct SenderGuard(pub Sender<Task>);
+
+impl std::ops::Drop for SenderGuard {
+    fn drop(&mut self) {
+        let _ = self.0.send(Task::Close);
+    }
+}
+
+enum Task {
+    Request { req: Request, result_tx: Sender<Response> },
+    Close,
+}
+
+struct Process {
+    path: PathBuf,
+    child: Child,
+}
+
+impl Process {
+    fn run(process_path: &Path) -> Result<Process, io::Error> {
+        let child = Command::new(process_path.clone())
+            .stdin(Stdio::piped())
```

Comment:
Let's also `.stderr(Stdio::null())`, lest a macro hijacks our stderr.

Reply:
Maybe we should capture stderr for debugging :grimacing:

Reply:
That would be perfect, but that again requires draining stdout and stderr in parallel, to avoid the problem where stderr overflows, and the proc-macro is blocked.

Reply:
Fixed.

Reply:
Hm, it is ifxed only in restart, and not here?